### PR TITLE
Refactor check controller

### PR DESCRIPTION
**Pull request by ChrisBr**
Supersedes https://github.com/openSUSE/open-build-service/pull/5796

* Implemented the suggested routing structure
* Introduced Status::Report model and dropped RepositoryPublish
* Implemented required_checks for projects

### CODE REVIEW

File: src/api/app/policies/status/report_policy.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  def show?
+    index?
+  end
+
+  private
+
+  def write_permission_for_checkable?(container)
+    container &&
```
Comment (Ana06) [inline src/api/app/policies/status/report_policy.rb]: I would do `return false unless container` and then the `||`. Much easier to read and you don't have the problem of splitting the lines (at least for the first one)

File: src/api/app/policies/status/report_policy.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      (container.relationships.users.pluck(:user_id).include?(@user.id) ||
+      GroupsUser.where(group: container.relationships.groups.pluck(:group_id), user: @user).exists?)
+  end
+
+  def request_target_object(action)
+    if action.target_package
```
Comment (Ana06) [inline src/api/app/policies/status/report_policy.rb]: This code is repeated in https://github.com/openSUSE/open-build-service/pull/5862/commits/2d89be4ccaed5d27948c22e6c84cde4727b6d51d#diff-66e753326c6844bc8ae62de2dfe0cb35R44.... :sob:

File: src/api/spec/controllers/status/reports_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  let(:project) { create(:project_with_repository) }
+  let(:source_project) { create(:project_with_package) }
+  let(:target_project) { create(:project_with_package) }
+  let(:bs_request) do
+    create(:bs_request_with_submit_action,
+           creator: user,
+           source_project: source_project,
+           source_package: source_project.packages.first,
+           target_project: target_project,
+           target_package: target_project.packages.first)
```
Comment (Ana06) [inline src/api/spec/controllers/status/reports_controller_spec.rb]: `target_project.packages.first` is repeated, maybe assign it to a variable? :thinking:

File: src/api/spec/models/status/report_spec.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      let(:bs_request) do
+        create(:bs_request_with_submit_action,
+               source_project: source_project,
+               source_package: source_project.packages.first,
+               target_project: target_project,
+               target_package: target_project.packages.first)
+      end
+      let(:status_report) { create(:status_report, checkable: bs_request) }
+
+      it { expect(status_report).not_to validate_presence_of(:uuid) }
```
Comment (Ana06) [inline src/api/spec/models/status/report_spec.rb]: is this needed? :thinking: I mean, you can test that it is valid.

File: src/api/app/controllers/status/reports_controller.rb [hunk 5]
```diff
@@ -49,7 +49,9 @@ def require_checkable
   end
 
   def xml_hash
```
Comment (Ana06) [inline src/api/app/controllers/status/reports_controller.rb]: where is this method used? :thinking:

File: src/api/spec/controllers/status/checks_controller_spec.rb [hunk 6]
```diff
@@ -169,8 +168,19 @@
...
     end
 
+    context 'with xml with empty element' do
+      let!(:relationship) { create(:relationship_project_user, user: user, project: project) }
+      let(:xml_with_empty_field) { '<check><short_description/></check>' }
+
+      subject! { put :update, body: xml_with_empty_field, params: { report_uuid: status_report.uuid, id: check.id }, format: :xml }
+
+      it { is_expected.to have_http_status(:success) }
+      it { expect(check.reload.short_description).to eq('') }
```
Comment (Ana06) [inline src/api/spec/controllers/status/checks_controller_spec.rb]: `to be_empty`

File: src/api/app/controllers/status/checks_controller.rb [hunk 7]
```diff
@@ -103,6 +62,7 @@ def set_xml_check
   def xml_hash
     result = HashWithIndifferentAccess.new
     parsed_body = Xmlhash.parse(request.body.read)
+    return if parsed_body.blank?
```
Comment (Ana06) [inline src/api/app/controllers/status/checks_controller.rb]: how is this related to the actions drop of this commit? :confused:

File: src/api/app/policies/project_policy.rb [hunk 8]
```diff
@@ -16,10 +24,21 @@ def destroy?
     update?
   end
 
+  def index?
+    true
+  end
+
+  def show?
+    index?
```
Comment (Ana06) [inline src/api/app/policies/project_policy.rb]: why not directly true? :confused:

**Comment by ChrisBr** [comment]
> What a huge PR! scream Is it possible to split it? It is really difficult to review it... tired_face Making changes suggested during the review is also going to be a nightmare.

Not really. This is a whole features, the different parts do not make sense separated...

> Why don't we need the other actions in 0c82948? Why is uuid_type not needed in 4a62829? It would be nice to add it in the commits 

I picked up the PR from @bgeuken who started in https://github.com/openSUSE/open-build-service/pull/5796. In this PR it was decided that we only need an :update and :index action. As this is not merged yet, I would suggest to just squash most of the commits together because that would answer probably some questions like

* why is uuid_type not needed 'anymore' (it was never needed)
* why are the other actions not needed 'anymore' (they were never needed)
* etc

File: src/api/app/controllers/status/reports_controller.rb [hunk 9]
```diff
@@ -1,41 +1,22 @@
 class Status::ReportsController < ApplicationController
-  before_action :set_status_report, only: [:show, :destroy]
   before_action :require_checkable
-  skip_before_action :require_login, only: [:show, :index]
-  after_action :verify_authorized, except: [:show, :index]
+  before_action :set_status_report
+  skip_before_action :require_login
```
Comment (hennevogel) [inline src/api/app/controllers/status/reports_controller.rb]: Please limit this to `:show`, otherwise the next developer will add another action that will then be accessible to everybody without him knowing...

**Comment by hennevogel** [comment]
I find the worst part of reviewing this is that you have not squashed the commits. Which makes it impossible to review by commit...

File: src/api/spec/models/status/report_spec.rb (hunk 4 above)
Comment (bgeuken) [reply src/api/spec/models/status/report_spec.rb]: This validation should only run when the checkable is a Repository, see https://github.com/openSUSE/open-build-service/pull/5862/files#diff-b70f033bbb90cc314855c93d00984b60R20

File: src/api/app/controllers/status/checks_controller.rb [hunk 10]
```diff
@@ -1,71 +1,58 @@
...
-      render_error(status: 422, errorcode: 'invalid_check', message: "Could not delete check: #{@check.errors.full_messages.to_sentence}")
+      @status_report = @checkable.status_reports.first_or_initialize
     end
   end
 
-  private
+  def require_checkable
+    set_repository || set_bs_request
+    return if @checkable
+    raise ActiveRecord::RecordNotFound
```
Comment (hennevogel) [inline src/api/app/controllers/status/checks_controller.rb]: As a User I don't know what is missing. So I can't fix my API call. What is the problem with having a chain of `before_actions`?

File: src/api/app/controllers/status/reports_controller.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+  private
+
+  def set_status_report
+    if params[:uuid]
+      @status_report = @checkable.status_reports.find_by(uuid: params[:uuid])
+    else
+      @status_report = @checkable.status_reports.first
+    end
+    raise ActiveRecord::RecordNotFound unless @status_report
```
Comment (hennevogel) [inline src/api/app/controllers/status/reports_controller.rb]: :thought_balloon:  Will this produce the right message?

File: src/api/app/controllers/status/reports_controller.rb [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  def require_checkable
+    set_repository || set_request
+    return if @checkable
+    raise ActiveRecord::RecordNotFound
```
Comment (hennevogel) [inline src/api/app/controllers/status/reports_controller.rb]: Same here as in the other controller...

File: src/api/app/controllers/status/checks_controller.rb [hunk 13]
```diff
@@ -74,17 +61,12 @@ def set_xml_check
...
-  def require_project
-    @project = Project.get_by_name(params[:project_name])
-  end
-
-  def require_repository
-    @repository = @project.repositories.find_by(name: params[:repository_name])
-    raise UnknownRepository, "Repository does not exist #{params[:repository_name]}" unless @repository
-  end
-
   def xml_hash
```
Comment (hennevogel) [inline src/api/app/controllers/status/checks_controller.rb]: This is private right?

File: src/api/app/controllers/status/reports_controller.rb [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    project = Project.find_by(name: params[:project_name])
+    @checkable = project.repositories.find_by(name: params[:repository_name]) if project
+  end
+
+  def set_request
+    return unless params[:bs_request_number]
+    @checkable = BsRequest.with_submit_requests.find_by(number: params[:bs_request_number])
+  end
+
+  def xml_hash
```
Comment (hennevogel) [inline src/api/app/controllers/status/reports_controller.rb]: Private...

File: src/api/app/controllers/status/required_checks_controller.rb [hunk 15]
```diff
@@ -63,20 +66,10 @@ def set_checkable
...
-      )
     else
-      render_error(
-        status: 404,
-        errorcode: 'not_found',
-        message: 'No repository name specified.'
-      )
+      @checkable = Project.get_by_name(params[:project_name])
     end
+    raise ActiveRecord::RecordNotFound unless @checkable
```
Comment (hennevogel) [inline src/api/app/controllers/status/required_checks_controller.rb]: :thought_balloon: Is this producing the right message?

File: src/api/app/models/repository.rb [hunk 16]
```diff
@@ -17,13 +17,13 @@ class Repository < ApplicationRecord
   has_many :repository_architectures, -> { order('position') }, dependent: :destroy, inverse_of: :repository
   has_many :architectures, -> { order('position') }, through: :repository_architectures
   has_many :status_publishes, class_name: 'Status::RepositoryPublish'
-  has_many :checks, through: :status_publishes, class_name: 'Status::Check' do
+  has_many :status_reports, as: :checkable, class_name: 'Status::Report', dependent: :destroy do
     def for_build_id(build_id)
```
Comment (hennevogel) [inline src/api/app/models/repository.rb]: :thought_balloon: This is `for_uuid` now right?

File: src/api/app/policies/status/report_policy.rb [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+    case @record.checkable
+    when BsRequest
+      @record.checkable.bs_request_actions.map(&:target_project_object)
+    when Repository
+      [@record.checkable.project]
+    end
+  end
+
+  def write_permission_for_checkable?(container)
+    container &&
```
Comment (hennevogel) [inline src/api/app/policies/status/report_policy.rb]: Why not use a guard clause?

File: src/api/app/policies/status/report_policy.rb [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+      (container.relationships.users.pluck(:user_id).include?(@user.id) ||
+      GroupsUser.where(group: container.relationships.groups.pluck(:group_id), user: @user).exists?)
```
Comment (hennevogel) [inline src/api/app/policies/status/report_policy.rb]: :thought_balloon: I really think this should be something generic in our objects that have user/group relationships. We keep doing this (`Object has user? || Object has a group with user`) in a gazillion places... But this is out of scope for this PR I guess...

File: src/api/db/migrate/20180906142702_create_status_reports.rb [hunk 19]
```diff
@@ -0,0 +1,16 @@
+class CreateStatusReports < ActiveRecord::Migration[5.2]
+  def change
+    create_table :status_reports, id: :integer, options: 'ROW_FORMAT=DYNAMIC' do |t|
+      t.string :uuid
+      t.belongs_to :checkable, polymorphic: { limit: 191 }, type: :integer
```
Comment (hennevogel) [inline src/api/db/migrate/20180906142702_create_status_reports.rb]: What is the limit for?

**Comment by Ana06** [comment]
> > What a huge PR! scream Is it possible to split it? It is really difficult to review it... tired_face Making changes suggested during the review is also going to be a nightmare.
> 
> Not really. This is a whole feature, the different parts do not make sense separated...

Some of the commits are refactoring that I think that could be separated. But maybe that changes once they are squashed.

> > Why don't we need the other actions in 0c82948? Why is uuid_type not needed in 4a62829? It would be nice to add it in the commits 
> 
> I picked up the PR from @bgeuken who started in https://github.com/openSUSE/open-build-service/pull/5796. In this PR it was decided that we only need an :update and :index action. As this is not merged yet, I would suggest to just squash most of the commits together because that would answer probably some questions like

Yes please! :pray: I think squashing the commits (and maybe adding some more info to the commit description) would be really helpful for the review. :wink:

File: src/api/spec/models/status/report_spec.rb (hunk 4 above)
Comment (Ana06) [reply src/api/spec/models/status/report_spec.rb]: I know that! :stuck_out_tongue_winking_eye:  I meant that you could test that an status report is valid without `uuid` when it is a BsRequest instead of testing `not_to validate_presence_of`.

File: src/api/app/controllers/status/checks_controller.rb [hunk 20]
```diff
@@ -8,9 +8,9 @@ class Status::ChecksController < ApplicationController
 
   # GET /projects/:project_name/repositories/:repository_name/repository_publishes/:repository_publish_build_id/checks
   def index
-    @checks = @checkable.checks
-    @missing_checks = @checkable.missing_checks
-    authorize @checks
+    authorize @status_report
```
[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/app/controllers/status/checks_controller.rb]: yes... but the `require_status_report` before action is defined in a commit which as after the commit where it is used :see_no_evil:

File: src/api/app/controllers/status/checks_controller.rb [hunk 21]
```diff
@@ -59,8 +59,13 @@ def require_or_initialize_checkable
   end
 
   def require_checkable
-    @checkable = Status::RepositoryPublish.find_by(build_id: params[:repository_publish_build_id]) if params[:repository_publish_build_id]
-    render_error(status: 404, errorcode: 'not_found', message: "Unable to find status_repository_publish with id '#{params[:repository_publish_build_id]}'") unless @checkable
+    if params[:repository_publish_build_id]
```
[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/app/controllers/status/checks_controller.rb]: Yes, but I think it is still not needed. `find_by` returns `nil` if you give it `nil` as a parameter. So this would do exactly the same if the `if` is not there. Conceptually also keeps the same.

File: src/api/app/policies/status/check_policy.rb [hunk 22]
```diff
@@ -7,9 +7,18 @@ def initialize(user, record)
...
+    case @record.checkable
+    when BsRequest
+      checkable_container = request_target_object
+    when Status::RepositoryPublish
+      checkable_container = @record.checkable
+    end
+
+    return false unless checkable_container
+
+    checkable_container.relationships.users.pluck(:user_id).include?(@user.id) || checkable_container.groups_users.pluck(:user_id).include?(@user.id)
```
[the remark this answers is not in the document]
Comment (ChrisBr) [reply src/api/app/policies/status/check_policy.rb]: The containers are always projects and we already have a ``can_modify`` method in project. Good catch @hennevogel :+1:

File: src/api/spec/controllers/status/reports_controller_spec.rb (hunk 3 above)
Comment (ChrisBr) [reply src/api/spec/controllers/status/reports_controller_spec.rb]: No, ``target_project.packages.first`` vs ``source_project.packages.first``

File: src/api/app/controllers/status/reports_controller.rb (hunk 5 above)
Comment (ChrisBr) [reply src/api/app/controllers/status/reports_controller.rb]: Not used anymore, thanks, good catch :+1:

File: src/api/app/controllers/status/checks_controller.rb [hunk 23]
```diff
@@ -52,17 +52,30 @@ def destroy
...
+    elsif params[:bs_request_number]
+      @checkable = BsRequest.find(params[:bs_request_number])
+    end
+
+    raise NotFoundError, "Couldn't find any checkable object." unless @checkable
+  end
+
   def require_checkable
     if params[:repository_publish_build_id]
       @checkable = Status::RepositoryPublish.find_by(build_id: params[:repository_publish_build_id])
```
[the remark this answers is not in the document]
Comment (ChrisBr) [reply src/api/app/controllers/status/checks_controller.rb]: Don't understand ...?
We don't have a uuid when it is a request if you mean that?

File: src/api/app/models/bs_request.rb [hunk 24]
```diff
@@ -88,6 +86,7 @@ class SaveError < APIError
   has_many :request_history_elements, -> { order(:created_at) }, class_name: 'HistoryElement::Request', foreign_key: :op_object_id
   has_many :review_history_elements, through: :reviews, source: :history_elements
   has_many :status_reports, as: :checkable, class_name: 'Status::Report', dependent: :destroy
+  has_many :target_project_objects, through: :bs_request_actions
```
[the remark this answers is not in the document]
Comment (ChrisBr) [reply src/api/app/models/bs_request.rb]: Associations yes but projects are not always project objects but strings :smile: The problem here is: In BsRequestAction there is an attribute target_project therefore we introduced the target_project_object association later on :scream: (You remember?). So this is to be 

* compliant with BsRequestAction
* otherwise we would need to specify the association name on the declaration of this association

File: src/api/db/migrate/20180906142702_create_status_reports.rb [hunk 25]
```diff
@@ -2,8 +2,7 @@ class CreateStatusReports < ActiveRecord::Migration[5.2]
   def change
     create_table :status_reports, id: :integer, options: 'ROW_FORMAT=DYNAMIC' do |t|
       t.string :uuid
-      t.integer :checkable_id
-      t.string :checkable_type
+      t.belongs_to :checkable, polymorphic: { limit: 191 }, type: :integer
```
[the remark this answers is not in the document]
Comment (ChrisBr) [reply src/api/db/migrate/20180906142702_create_status_reports.rb]: I had a weird issue that this migration failed with the new utf8mb4 encoding which the limit fixes (number of bytes allowed IIRC). Can you check if this works for you without the limit?

[the remark this answers is not in the document]
Comment (coolo) [reply src/api/db/migrate/20180906142702_create_status_reports.rb]: did you already have ROW_FORMAT=DYNAMIC in place when you had the problem? Because that's the real fix.

[the remark this answers is not in the document]
Comment (ChrisBr) [reply src/api/db/migrate/20180906142702_create_status_reports.rb]: @coolo IIRC yup that was already there but could be I messed sth up with the migrations. Anyway , would be good if sb can double check it!

File: src/api/app/controllers/status/checks_controller.rb (hunk 20 above)
[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/app/controllers/status/checks_controller.rb]: with the squashing this is solved :wink:

File: src/api/db/migrate/20180906142702_create_status_reports.rb (hunk 19 above)
Comment (hennevogel) [reply src/api/db/migrate/20180906142702_create_status_reports.rb]: What @coolo comment?

Comment (coolo) [reply src/api/db/migrate/20180906142702_create_status_reports.rb]: But if I get this right, you remove it in a later migration anyway - so just drop all of it? :)

File: src/api/spec/models/status/report_spec.rb (hunk 4 above)
Comment (hennevogel) [reply src/api/spec/models/status/report_spec.rb]: You say tomatoes

@Ana06 you need to start tagging your nitpicks like we have agreed we would do. Please :smiley:

File: src/api/db/migrate/20180906142702_create_status_reports.rb (hunk 19 above)
Comment (hennevogel) [reply src/api/db/migrate/20180906142702_create_status_reports.rb]: No we do not drop it the migration. This is what is going to happen.

Comment (coolo) [reply src/api/db/migrate/20180906142702_create_status_reports.rb]: the 3 migrations are just within this PR, this is just squashing them.

File: src/api/app/models/status/report.rb [hunk 26]
```diff
@@ -0,0 +1,45 @@
+class Status::Report < ApplicationRecord
+  #### Includes and extends
```
[the remark this answers is not in the document]
Comment (hennevogel) [reply src/api/app/models/status/report.rb]: This is the template we came up with. Change them, they reside in `lib/templates` :cowboy_hat_face:

**Comment by hennevogel** [comment]
No idea how to address the remaining codeclimate suggestion. I would mark it as wontfix.

File: src/api/db/migrate/20180906142702_create_status_reports.rb (hunk 25 above)
[the remark this answers is not in the document]
Comment (hennevogel) [reply src/api/db/migrate/20180906142702_create_status_reports.rb]: Yeah the structure.sql test in the package fails without explicitly setting this as it's DB version dependent. I just love this test...

File: src/api/db/migrate/20180906142702_create_status_reports.rb (hunk 19 above)
Comment (hennevogel) [reply src/api/db/migrate/20180906142702_create_status_reports.rb]: Yeah the migration test in the package fails without it as it's DB version dependent. :angel:  I :heart: this one...

File: src/api/app/controllers/status/reports_controller.rb [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+
+  # GET /status_reports/published/:project_name/:repository_name/reports/:uuid
+  def show
+    @checks = @status_report.checks
+    @missing_checks = @status_report.missing_checks
+  end
+
+  private
+
+  def set_status_report
```
Comment (DavidKang) [inline src/api/app/controllers/status/reports_controller.rb]: Shouldn't this be part of the concern? :thinking:

File: src/api/app/controllers/status/required_checks_controller.rb [hunk 28]
```diff
@@ -63,20 +66,10 @@ def set_checkable
     if params[:repository_name]
```
Comment (DavidKang) [inline src/api/app/controllers/status/required_checks_controller.rb]: Why not using here the concern? :thinking:

Comment (DavidKang) [reply src/api/app/controllers/status/required_checks_controller.rb]: @hennevogel, thanks for the clarification

**Comment by hennevogel** [comment]
deployed too. 

checks:
```
osc api /status_reports/published/home:hennevogel/openSUSE_Tumbleweed/reports/18
osc api -X PUT /status_reports/published/home:hennevogel/openSUSE_Tumbleweed/reports/18/checks -T check.xml
```

required checks:
```
osc api /status_reports/repositories/home:hennevogel/openSUSE_Tumbleweed/required_checks
osc api -X POST -T required_checks.xml /status_reports/repositories/home:hennevogel/openSUSE_Tumbleweed/required_checks
```

**Comment by hennevogel** [comment]
See #5886
